File: 2022/day03.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <set>
#include <map>

using namespace std;

set<char> intersection(string& s1, string& s2 ){
    set<char> inter;
    for(int i = 0; i < s1.size(); i++){
        for(int j = 0; j < s2.size(); j++){
            if(s1[i] == s2[j]){
                inter.insert(s1[i]);
            }
        }
    }
    return inter;
}

string set_to_string(set<char>& myset){
    set<char>::const_iterator setIterator;
    string s;
    for(setIterator = myset.begin(); setIterator != myset.end(); setIterator++) {
        s+= *setIterator;
    }
    return s;
}

void first_part(map<char, int>& priorities){
    string line;
    ifstream inFile("input/day03.txt");
    string s1, s2;
    string str;
    
    set<char> s;

    int priorities_sum = 0;

    while(getline(inFile, line)) { 
        s1 = "";
        s2 = "";
        for(int i = 0; i < line.size(); i++){
            if (i < line.size()/2){
                s1 += line[i];
            }
            else{
                s2 += line[i];
            }
        }
        s = intersection(s1,s2);
        str = set_to_string(s);

        for(int i = 0; i < str.size(); i++){
            priorities_sum += priorities[str[i]];
        }
    }
    // cout << s1 << endl; 
    // cout << s2 << endl; 
    
    // cout << str << endl; 
    
    cout << priorities_sum << endl; 
}

void second_part(map<char, int>& priorities){
    string line;
    ifstream inFile("input/day03.txt");
    
    string s1, s2, s3, str;
    set<char> s;

    int priorities_sum = 0;

    while(getline(inFile, line)) { 
        s1 = line;
        getline(inFile, line);
        s2 = line;

        s = intersection(s1,s2);
        str = set_to_string(s);

        getline(inFile, line);
        s3 = line;

        s = intersection(str,s3);
        str = set_to_string(s);

        for(int i = 0; i < str.size(); i++){
            priorities_sum += priorities[str[i]];
        }
    }
    cout << priorities_sum << endl;
    
}

int main() {
  string letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  map<char, int> priorities;
  for(int i=0; i<letters.size(); i++){
    priorities[letters[i]] = i+1;
    // cout << letters[i] << ": " << priorities[letters[i]] << endl;
  }

  // part 1
  first_part(priorities);

  // part 2
  second_part(priorities);
}

File: 2022/day03.py
letters = 'abcdefghijklmnopqrstuvwxyz'
letters = letters + ''.join([l.upper() for l in letters])
print(letters)
priorities = {l:i+1 for i, l in enumerate(letters)}

filename = "input/day03.txt"
# filename = "input/day03_sample.txt"

# part 1
priorities_sum = 0
with open(filename, 'r') as f:
    lines = f.readlines()
    for line in lines:
        
        mid = len(line) // 2
        left = set(line[:mid])
        right = set(line[mid:])
        for c in left.intersection(right):
            priorities_sum += priorities[c]

print(priorities_sum)

# part 2
priorities_sum = 0
with open(filename, 'r') as f:
    lines = f.readlines()
    i  = 0
    while i < len(lines):
        common = set(lines[i].strip()).intersection(set(lines[i+1].strip())).intersection(set(lines[i+2].strip()))
        for c in common:
            priorities_sum += priorities[c]
        i = i + 3
        
print(priorities_sum)
